docs/rfcs/1012-parsec-auth.md

```diff
@@ -0,0 +1,432 @@
+> The output of `PBKDF_A` and `PBKDF_B` should be different to avoid leaking the private key.
+
+
+```mermaid
+sequenceDiagram
+  autonumber
+  participant Alice
+  participant S as Authentification server
+
+  Alice ->> S: Provide its email for the account creation
+  S ->> Alice: Send an email with a unique code need to continue the process
+  alt Choose between password or fido2
+    Alice ->> Alice: Choose a password
+    Alice ->> Alice: Derive `AUTH_MEDIUM_HMAC_KEY` using the password and `PBKDF_A`
+    Alice ->> Alice: Create `AUTH_MEDIUM_SYM_KEY` using the password and `PBKDF_B`
+    Alice ->> Alice: Generate `AUTH_MEDIUM_PRIV_KEY`, `AUTH_MEDIUM_PUB_KEY`
+    Alice ->> Alice: Encrypt `AUTH_MEDIUM_PRIV_KEY` with `AUTH_MEDIUM_SYM_KEY`
+  else
+    Alice ->> Alice: Generate `AUTH_MEDIUM_HMAC_KEY` using the fido2 device.
+    Alice ->> Alice: Use the public key of the FIDO2 device as `AUTH_MEDIUM_PUB_KEY`
+  end
+
+  Alice ->> Alice: Create a bundle containing `AUTH_MEDIUM_PUB_KEY` (`ACCOUNT_MANIFEST`)
+  Alice ->> Alice: Create a symmetric key `ACCOUNT_MANIFEST_SYM_KEY`
+  Alice ->> Alice: Encrypt the bundle `ACCOUNT_MANIFEST` with `ACCOUNT_MANIFEST_SYM_KEY` (`enc(ACCOUNT_MANIFEST)`)
+  Alice ->> Alice: Encrypt `ACCOUNT_MANIFEST_SYM_KEY` with `AUTH_MEDIUM_PUB_KEY` (`enc(ACCOUNT_MANIFEST_SYM_KEY)`)
+  Alice ->> S: Send the require data to finish the account creation<br>(`enc(AUTH_MEDIUM_PRIV_KEY)`, `AUTH_MEDIUM_SYM_KEY_PARAMS`, `enc(ACCOUNT_MANIFEST)`,<br>`enc(ACCOUNT_MANIFEST_SYM_KEY)`, `AUTH_MEDIUM_HMAC_KEY`, email validation token)
+```
+
+At the end of the process, the server should save the following information:
+
+- An entry for the user account:
+
+  - A UUID (`account_id`)
+
+  - The email of the user
+
+  - The encrypted account manifest (`enc(ACCOUNT_MANIFEST)`)
+
+- An entry for the authentication method linked to the user:
+
+  - A UUID (`auth_medium_id`)
+  - The HMAC secret (`AUTH_MEDIUM_HMAC_KEY`)
+
+    The secret should be securely stored by either encrypting its value or using a vault.
+
+  - The encrypted private key (`enc(AUTH_MEDIUM_PRIV_KEY)`)
+
+    > Could be empty when the user uses a FIDO2 device.
+
+  - The encrypted key used to decrypt the account manifest (`enc(ACCOUNT_MANIFEST_SYM_KEY)`)
+  - The parameters used to generate the symmetric key (`AUTH_MEDIUM_SYM_KEY`) that encrypted `enc(AUTH_MEDIUM_PRIV_KEY)` (`AUTH_MEDIUM_SYM_KEY_PARAMS`)
+
+> [!WARNING]
+> A user should be able to register multiple authentication methods,
+
+### The authentication process
+
+We use a HMAC to authenticate each request to the server with the secret `AUTH_MEDIUM_HMAC_KEY` that was previously shared.
+
+### Obtain the account manifest from the server
+
+The server store the encrypted account manifest that contain the important information about the user.
+For the user to obtain the account manifest from the server it will do:
+
+1. Be authenticated with the server.
+2. Retrieve the following encrypted data from the sever:
```

Comment:
Also note that since we authenticate the request with a given auth medium id, we should only be able to get back the data related to this specific auth medium.

Reply:
I explicitly indicated that the encrypted blob correspond to the current authentication method